lib/stream.js

```diff
@@ -105,6 +105,29 @@ const cleanupStdioStreams = (customStreams, error) => {
 	}
 };
 
+// Like `once()` except it never rejects, especially not on `error` event.
+const pEvent = (eventEmitter, eventName) => new Promise(resolve => {
+	eventEmitter.once(eventName, (...payload) => {
+		resolve([eventName, ...payload]);
+	});
+});
+
+const throwOnProcessError = async processErrorPromise => {
+	const [, error] = await processErrorPromise;
+	throw error;
+};
+
+// First the `spawn` event is emitted, then `exit`.
+// If the `error` event is emitted:
+//  - before `spawn`: `exit` is never emitted.
+//  - after `spawn`: `exit` is always emitted.
+// We only want to listen to `exit` if it will be emitted, i.e. if `spawn` has been emitted.
+// Therefore, the arguments order of `Promise.race()` is significant.
+const waitForFailedProcess = async (processSpawnPromise, processErrorPromise, processExitPromise) => {
+	const [eventName] = await Promise.race([processSpawnPromise, processErrorPromise]);
```

Comment:
Either `spawn` or `error` is emitted.
Node.js source code about this: https://github.com/nodejs/node/blob/0056147a87c3964f83cb3c4b31a7b7fbfcba28e8/lib/internal/child_process.js#L398-L424

That's also [documented](https://nodejs.org/api/child_process.html#event-spawn):

> If the child process does not spawn successfully, the 'spawn' event is not emitted and the 'error' event is emitted instead.